auth/authenticator.go

```diff
@@ -0,0 +1,225 @@
+// © Copyright IBM Corporation 2020. All Rights Reserved.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at:
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package auth
+
+import (
+	"bytes"
+	"fmt"
+	"net/http"
+	"strings"
+	"sync"
+	"time"
+
+	"github.com/IBM/go-sdk-core/v4/core"
+)
+
+const (
+	AUTHTYPE_COUCHDB_SESSION = "couchdbSession"
+)
+
+var requestSessionMutex sync.Mutex
+
+// CouchDbSessionAuthenticator uses username and password to obtain
+// CouchDB authentication cookie, and adds the cookie to requests.
+type CouchDbSessionAuthenticator struct {
+	URL, Username, Password string
+	Client                  *http.Client
+	session                 *session
+}
+
+// GetAuthenticatorFromEnvironment instantiates an Authenticator
+// using service properties retrieved from external config sources.
+func GetAuthenticatorFromEnvironment(credentialKey string) (authenticator core.Authenticator, err error) {
+	props, err := core.GetServiceProperties(credentialKey)
+	if err != nil {
+		return
+	}
+	if props[core.PROPNAME_AUTH_TYPE] == "COUCHDB_SESSION" {
+		authenticator, err = NewCouchDbSessionAuthenticatorFromMap(props)
+		return
+	}
+
+	return core.GetAuthenticatorFromEnvironment(credentialKey)
+}
+
+// NewCouchDbSessionAuthenticator constructs a new NewCouchDbSessionAuthenticator instance.
+func NewCouchDbSessionAuthenticator(url, username, password string) (*CouchDbSessionAuthenticator, error) {
+	authenticator := &CouchDbSessionAuthenticator{
+		URL:      url,
+		Username: username,
+		Password: password,
+	}
+	if err := authenticator.Validate(); err != nil {
+		return nil, err
+	}
+	return authenticator, nil
+}
+
+func NewCouchDbSessionAuthenticatorFromMap(props map[string]string) (*CouchDbSessionAuthenticator, error) {
+	if props == nil {
+		return nil, fmt.Errorf(core.ERRORMSG_PROPS_MAP_NIL)
+	}
+	url := props[core.PROPNAME_SVC_URL]
+	username := props[core.PROPNAME_USERNAME]
+	password := props[core.PROPNAME_PASSWORD]
+	return NewCouchDbSessionAuthenticator(url, username, password)
+}
+
+// AuthenticationType returns the authentication type for this authenticator.
+func (a CouchDbSessionAuthenticator) AuthenticationType() string {
+	return AUTHTYPE_COUCHDB_SESSION
+}
+
+// Validate the authenticator's configuration.
+// Ensures the service url, username and password are valid and not nil.
+func (a CouchDbSessionAuthenticator) Validate() error {
```

Comment:
This is against golang's style conventions though, using a one-letter variable for a receiver is extremely common and is generally recommended in [Go Code Review Comments](https://github.com/golang/go/wiki/CodeReviewComments#variable-names). The reasoning goes like:

> The greater the distance between a name's declaration and its uses,
the longer the name should be. 

Ofc we can ignore go's convention, but I'm not sure how beneficial is it - we can't use `authenticator` as a var name, because this is a name of the package and I don't feel that `auth` makes things that much more readable, given that most of the functions is ~8 lines long and var's descriptive type is right here in the method definition.

Reply:
Ok interesting, Go seems to be the opposite compared to other language styles that suggest using meaningful variable names.

Reply:
Yes, golang's naming is rather controversial, influenced by Rob Pike's (one of the creators) philosophy, as far as I understand. There are a couple presentations on the topic around, e.g. [https://talks.golang.org/2014/names.slide#1](https://talks.golang.org/2014/names.slide#1)

Personally this felt rather strange when I was starting with the language, but now it's other way around, for example I find `core`'s style is very unnatural.

That said I still can switch from `a` to `auth` if you think it'll really increase readability and not just a question of naming preference coming from other languages.

Reply:
No need to change it, let's stick with the naming convention.

Reply:
@eiri What do you think about the variable `this`? As I can see this is used also in the core: https://github.com/IBM/go-sdk-core/blob/775201f679d69afb97f3259ababb24d482089410/v4/core/basic_authenticator.go#L76

Reply:
@vmatyusGitHub I dislike javaisms in go code, I think it's misleading to bring in primitives from other languages, especially from the languages with a different paradigm.

As I said, `core`'s style looks very unnatural to me and I'd prefer not to mimic it, unless it's really necessary for users familiarity.

Reply:
Okay, I can accept this point of view. Just was an idea.